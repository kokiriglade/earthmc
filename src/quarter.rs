//! # Quarter
//!
//! Defines the [`Quarter`] struct.
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    block_position::BlockPosition,
    named_id::{NamedId, NamedIdOpt},
};

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Quarter {
    /// The Quarter's name.
    pub name: String,
    /// The Quarter's [`Uuid`].
    pub uuid: Uuid,
    /// The Quarter's kind, or as it is referred to in the API, its `type`.
    #[serde(rename = "type")]
    pub kind: QuarterKind,
    /// The [`Uuid`] of the creator of the Quarter.
    pub creator: Uuid,
    /// The Quarter's owner.
    pub owner: NamedIdOpt,
    /// The town this Quarter is inside.
    pub town: NamedId,
    pub timestamps: QuarterTimestamps,
    pub status: QuarterStatus,
    pub stats: QuarterStats,
    /// Tuple of RGBA values of the Quarter.
    pub colour: [u8; 4],
    /// All of the trusted players in this Quarter.
    pub trusted: Vec<NamedId>,
    /// Every cuboid in this Quarter.
    pub cuboids: Vec<QuarterCuboid>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum QuarterKind {
    /// The default kind of Quarter.
    Apartment,
    /// Allows bed usage.
    Inn,
    /// Allows vehicle placing and usage.
    Station,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QuarterTimestamps {
    /// Unix timestamp representing the time the Quarter was created.
    pub registered: i64,
    /// Unix timestamp representing the time the Quarter was claimed.
    pub claimed_at: Option<i64>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QuarterStatus {
    /// If the Quarter is an embassy.
    pub is_embassy: bool,
    /// If the Quarter is for sale.
    pub is_for_sale: bool,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QuarterStats {
    /// The Quarter's sale price.
    pub price: Option<i32>,
    /// The total number of blocks within the Quarter's bounds.
    pub volume: i32,
    /// The total amount of cuboids this Quarter is made of.
    pub num_cuboids: u16,
    /// The size of the particles displayed in-game.
    pub particle_size: Option<f32>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QuarterCuboid {
    pub corner_one: BlockPosition,
    pub corner_two: BlockPosition,
}
