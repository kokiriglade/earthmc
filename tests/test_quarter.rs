use earthmc::quarter::Quarter;

#[test]
fn test_quarter() {
    let raw_json = include_str!("inputs/quarter.json");
    let parsed: Vec<Quarter> = serde_json::from_str(raw_json).unwrap();
    insta::assert_debug_snapshot!(parsed);
}
