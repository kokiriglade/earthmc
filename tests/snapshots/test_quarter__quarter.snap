---
source: tests/test_quarter.rs
expression: parsed
---
[
    Quarter {
        name: "PKTUI Room 2",
        uuid: 971c5315-12c2-4c15-a8c3-d65121fa4e07,
        kind: Apartment,
        creator: 766ef3ed-a9f9-4a57-8a75-27ca91244fe6,
        owner: NamedIdGeneric {
            name: Some(
                "Hungry482",
            ),
            uuid: Some(
                8bcbbb9e-074e-421d-9ee4-6cd13326bddc,
            ),
        },
        town: NamedIdGeneric {
            name: "Dfam",
            uuid: 4dccac1a-9b4a-4a70-90dc-5ce3744806d7,
        },
        timestamps: QuarterTimestamps {
            registered: 1746089983782,
            claimed_at: Some(
                1746090794781,
            ),
        },
        status: QuarterStatus {
            is_embassy: false,
            is_for_sale: false,
        },
        stats: QuarterStats {
            price: None,
            volume: 336,
            num_cuboids: 1,
            particle_size: None,
        },
        colour: [
            63,
            180,
            255,
            255,
        ],
        trusted: [],
        cuboids: [
            QuarterCuboid {
                corner_one: [
                    7597,
                    70,
                    -3219,
                ],
                corner_two: [
                    7584,
                    67,
                    -3224,
                ],
            },
        ],
    },
]
